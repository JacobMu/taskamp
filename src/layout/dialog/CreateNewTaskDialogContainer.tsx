import React from "react";
import moment from "moment";

import {
  Dialog,
  FormGroup,
  InputGroup,
  TextArea,
  Label,
  Button,
  ButtonGroup,
  Divider,
} from "@blueprintjs/core";
import { DateInput } from "@blueprintjs/datetime";

import { InputErrorView } from "./InputErrorView";
import { useCreateNewTaskDialog } from "./useCreateNewTaskDialog";
import {
  FORM_FIELD_NAME,
  TaskProps,
  TASK_PRIORITY,
  DATE_FORMAT,
} from "./CreateNewTaskDialogService";

import "./CreateNewTaskDialogView.scss";

const SELECT_ITEMS = [
  { value: TASK_PRIORITY.LOWEST },
  { value: TASK_PRIORITY.LOW },
  { value: TASK_PRIORITY.MEDIUM },
  { value: TASK_PRIORITY.HIGH },
  { value: TASK_PRIORITY.HIGHEST },
];

interface Props {
  isDialogOpen: boolean;
  handleClose(): void;
  dispatchNewCard(cardProps: TaskProps): void;
}

export const CreateNewTaskDialogContainer: React.FC<Props> = ({
  isDialogOpen,
  handleClose,
  dispatchNewCard,
}) => {
  const {
    handleConfirm,
    handleInput,
    formErrors,
    isTaskSubmitted,
    creationTime,
  } = useCreateNewTaskDialog(dispatchNewCard);

  const showDialog = isTaskSubmitted ? false : isDialogOpen;

  return (
    <Dialog
      isOpen={showDialog}
      isCloseButtonShown
      canOutsideClickClose
      onClose={handleClose}
      title="createNewTaskDialog.title"
    >
      <div className="formGroupPadding">
        <FormGroup>
          <Label>
            {"dialog.form.taskId.label"}
            <InputGroup
              fill
              onChange={(event: React.ChangeEvent<HTMLInputElement>) =>
                handleInput(FORM_FIELD_NAME.TASK_ID, event.target.value)
              }
            />
            {formErrors?.TASK_ID.hasError && <InputErrorView />}
          </Label>
          <Label>
            {"dialog.form.taskNameLabel"}
            <InputGroup
              fill
              onChange={(event: React.ChangeEvent<HTMLInputElement>) =>
                handleInput(FORM_FIELD_NAME.TASK_NAME, event.target.value)
              }
            />
            {formErrors?.TASK_NAME.hasError && <InputErrorView />}
          </Label>
          <Label>
            {"dialog.form.dateLabel"}
            <DateInput
              fill
              formatDate={(date) => moment(date).format(DATE_FORMAT)}
              defaultValue={creationTime}
              disabled
            />
          </Label>
          <Label>
            {"dialog.form.describtionLabel"}
            <TextArea
              fill
              onChange={(event) =>
                handleInput(
                  FORM_FIELD_NAME.TASK_DESCRIPTION,
                  event.target.value
                )
              }
            />
            {formErrors?.TASK_DESCRIPTION.hasError && <InputErrorView />}
          </Label>
          <Label>
            {"dialog.form.selectLabel"}
            <div className="bp3-select">
              <select
                onChange={(event) =>
                  handleInput(FORM_FIELD_NAME.TASK_PRIO, event.target.value)
                }
              >
                {SELECT_ITEMS.map((item) => (
                  <option key={item.value} value={item.value}>
                    {item.value}
                  </option>
                ))}
              </select>
            </div>
          </Label>
        </FormGroup>
        <Divider />
        <ButtonGroup className="formGroupPadding">
          <Button onClick={handleConfirm}>{"dialog.confirmBtn"}</Button>
        </ButtonGroup>
      </div>
    </Dialog>
  );
};
