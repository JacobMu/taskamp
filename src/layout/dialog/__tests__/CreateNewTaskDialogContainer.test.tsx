import React from "react";
import { shallow } from "enzyme";

import { CreateNewTaskDialogContainer } from "../CreateNewTaskDialogContainer";

jest.mock("../useCreateNewTaskDialog", () => ({
  useCreateNewTaskDialog: jest.fn().mockReturnValue({
    handleConfirm: jest.fn().mockName("handleConfirmMock"),
    handleInput: () => {},
    formErrors: undefined,
    isTaskSubmitted: false,
    creationTime: ("12.12.2020" as unknown) as Date,
  }),
}));

const DEFAULT_PROPS = {
  isDialogOpen: true,
  handleClose: jest.fn().mockName("handleCloseMock"),
  dispatchNewCard: () => {},
};
describe("<CreateNewTaskDialogContainer />", () => {
  it("renders dialog with form", () => {
    const wrapper = shallow(<CreateNewTaskDialogContainer {...DEFAULT_PROPS} />);

    expect(wrapper).toMatchSnapshot();
  });
});
