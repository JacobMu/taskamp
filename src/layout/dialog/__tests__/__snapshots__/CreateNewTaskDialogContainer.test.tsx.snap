// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CreateNewTaskDialogContainer /> renders dialog with form 1`] = `
<Blueprint3.Dialog
  canOutsideClickClose={true}
  isCloseButtonShown={true}
  isOpen={true}
  onClose={[MockFunction handleCloseMock]}
  title="createNewTaskDialog.title"
>
  <div
    className="formGroupPadding"
  >
    <Blueprint3.FormGroup>
      <Component>
        dialog.form.taskId.label
        <Blueprint3.InputGroup
          fill={true}
          onChange={[Function]}
        />
      </Component>
      <Component>
        dialog.form.taskNameLabel
        <Blueprint3.InputGroup
          fill={true}
          onChange={[Function]}
        />
      </Component>
      <Component>
        dialog.form.dateLabel
        <Blueprint3.DateInput
          closeOnSelection={true}
          dayPickerProps={Object {}}
          defaultValue="12.12.2020"
          disabled={true}
          fill={true}
          formatDate={[Function]}
          invalidDateMessage="Invalid date"
          maxDate={2020-12-31T15:42:44.608Z}
          minDate={2000-01-01T15:42:44.608Z}
          outOfRangeMessage="Out of range"
          reverseMonthAndYearMenus={false}
        />
      </Component>
      <Component>
        dialog.form.describtionLabel
        <Blueprint3.TextArea
          fill={true}
          onChange={[Function]}
        />
      </Component>
      <Component>
        dialog.form.selectLabel
        <div
          className="bp3-select"
        >
          <select
            onChange={[Function]}
          >
            <option
              key="LOWEST"
              value="LOWEST"
            >
              LOWEST
            </option>
            <option
              key="LOW"
              value="LOW"
            >
              LOW
            </option>
            <option
              key="MEDIUM"
              value="MEDIUM"
            >
              MEDIUM
            </option>
            <option
              key="HIGH"
              value="HIGH"
            >
              HIGH
            </option>
            <option
              key="HIGHEST"
              value="HIGHEST"
            >
              HIGHEST
            </option>
          </select>
        </div>
      </Component>
    </Blueprint3.FormGroup>
    <Blueprint3.Divider />
    <Blueprint3.ButtonGroup
      className="formGroupPadding"
    >
      <Blueprint3.Button
        onClick={[MockFunction handleConfirmMock]}
      >
        dialog.confirmBtn
      </Blueprint3.Button>
    </Blueprint3.ButtonGroup>
  </div>
</Blueprint3.Dialog>
`;
